sunpy/coordinates/metaframes.py

```diff
@@ -0,0 +1,233 @@
+"""
+Coordinate frames that are defined relative to other frames
+"""
+
+from astropy import units as u
+from astropy.coordinates import SphericalRepresentation
+from astropy.coordinates.attributes import Attribute, QuantityAttribute
+from astropy.coordinates.baseframe import frame_transform_graph
+from astropy.coordinates.transformations import FunctionTransform
+
+from sunpy import log
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+from .frames import HeliocentricInertial, SunPyBaseCoordinateFrame
+from .offset_frame import NorthOffsetFrame
+from .transformations import _transformation_debug
+
+__all__ = ['NorthOffsetFrame', 'RotatedSunFrame']
+
+
+# The code for NorthOffsetFrame currently lives in `offset_frame.py`
+# This changes its module to this file so that the docs think that the class is local
+NorthOffsetFrame.__module__ = __name__
+
+
+_rotatedsun_cache = {}
+
+
+def _make_rotatedsun_cls(framecls):
+    """
+    Create a new class that is the rotated-Sun frame for a specific class of
+    base frame. If such a class has already been created for this frame, the
+    same class will be returned.
+
+    This function is necessary because frame transformations depend
+    on connection between specific frame *classes*.  So each type of frame
+    needs its own distinct rotated-Sun frame class.  This function generates
+    just that class, as well as ensuring that only one example of such a class
+    actually gets created in any given Python session.
+    """
+    # This code reuses significant code from Astropy's implementation of SkyOffsetFrame
+
+    if framecls in _rotatedsun_cache:
+        return _rotatedsun_cache[framecls]
+
+    # the class of a class object is the metaclass
+    framemeta = type(framecls)
+
+    class RotatedSunMeta(framemeta):
+        """
+        This metaclass renames the class to be "RotatedSun<framecls>".
+        """
+        def __new__(cls, name, bases, members):
+            # This has to be done because FrameMeta will set these attributes
```

Comment:
If this comment is to believed why does this not need to be done after the call to `super()`?

Reply:
I thought I understood this block (copied from `SkyOffsetFrame`), but now I'm not sure.  I'll do some testing.

Reply:
This code doesn't appear to be needed for the desired behavior (at least anymore), so may be vestigial in `SkyOffsetFrame`.  I'll remove it here in my next pass.